core/src/main/java/org/elasticsearch/action/support/master/TransportMasterNodeAction.java

```diff
@@ -48,6 +50,19 @@
  * A base class for operations that needs to be performed on the master node.
  */
 public abstract class TransportMasterNodeAction<Request extends MasterNodeRequest, Response extends ActionResponse> extends HandledTransportAction<Request, Response> {
+    private static final ClusterStateObserver.ChangePredicate masterNodeChangedPredicate = new ClusterStateObserver.ChangePredicate() {
+        @Override
+        public boolean apply(ClusterState previousState, ClusterState.ClusterStateStatus previousStatus,
+                             ClusterState newState, ClusterState.ClusterStateStatus newStatus) {
+            return newState.nodes().masterNodeId() != null
+                    && !newState.nodes().masterNodeId().equals(previousState.nodes().masterNodeId());
```

Comment:
can we change this to support the case where master election started and completed between the sampling of the cluster state in the observer and when this function gets called? I think `newState.nodes().masterNodeId() != null
                    && newState != previousState);` is good? (plus a comment)